File: backend/app/api/remote_plugin.py
"""远程插件管理 API 路由。"""
from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from ..deps import CurrentUser, DBSession
from ..schemas.remote_plugin import RemotePluginCreate, RemotePluginOut
from ..services import feature_service
from ..services import remote_plugin_service as svc
from ..services.remote_plugin_service import (
    DuplicatePluginName,
    GitOperationFailed,
    InvalidPluginMetadata,
    InvalidSourceUrl,
    RemotePluginError,
    RemotePluginNotFound,
)

router = APIRouter(prefix="/api/remote-plugins", tags=["remote-plugins"])


@router.get("", response_model=list[RemotePluginOut])
async def list_remote_plugins(db: DBSession, _user: CurrentUser):
    """列出所有已安装远程插件。"""
    rows = await svc.list_installed(db)
    return rows


@router.post("/install", response_model=RemotePluginOut, status_code=201)
async def api_install_plugin(
    body: RemotePluginCreate, db: DBSession, _user: CurrentUser
):
    """从 Git URL 克隆并安装远程插件。

    ``default_enabled=True`` 时，安装后自动为所有已有账号启用该插件
    （写入 AccountFeature 行），功能矩阵 Tab 会自动展示。
    """
    try:
        row = await svc.install(
            db, body.source_url, default_enabled=body.default_enabled,
        )
        await db.commit()
        await db.refresh(row)
        await svc.trigger_reload(db, row.name)
        return row
    except DuplicatePluginName as e:
        raise HTTPException(409, detail={"code": e.code, "message": e.message}) from e
    except InvalidSourceUrl as e:
        raise HTTPException(400, detail={"code": e.code, "message": e.message}) from e
    except (GitOperationFailed, InvalidPluginMetadata) as e:
        raise HTTPException(400, detail={"code": e.code, "message": e.message}) from e
    except RemotePluginError as e:
        raise HTTPException(400, detail={"code": e.code, "message": e.message}) from e
    except Exception as e:
        raise HTTPException(500, detail={"code": "INTERNAL", "message": f"安装失败: {e}"}) from e


@router.post("/{name}/enable")
async def api_enable(name: str, db: DBSession, _user: CurrentUser):
    """启用指定远程插件（全局开关）。"""
    try:
        row = await svc.enable(db, name, bootstrap_accounts=True)
        plugin_name = row.name
        await db.commit()
        await svc.trigger_reload(db, plugin_name)
        return {"ok": True, "name": plugin_name, "enabled": True}
    except RemotePluginNotFound as e:
        raise HTTPException(404, detail={"code": e.code, "message": e.message}) from e


@router.post("/{name}/disable")
async def api_disable(name: str, db: DBSession, _user: CurrentUser):
    """禁用指定远程插件（全局开关）。"""
    try:
        row = await svc.disable(db, name)
        plugin_name = row.name
        await db.commit()
        await svc.trigger_reload(db, plugin_name)
        return {"ok": True, "name": plugin_name, "enabled": False}
    except RemotePluginNotFound as e:
        raise HTTPException(404, detail={"code": e.code, "message": e.message}) from e


class AccountPluginAction(BaseModel):
    account_ids: list[int]


@router.post("/{name}/enable-accounts")
async def api_enable_accounts(name: str, body: AccountPluginAction, db: DBSession, _user: CurrentUser):
    """按账号启用远程插件（写入 AccountFeature 行，功能矩阵可见）。"""
    # 先确认远程插件存在
    rp = await svc.get_by_name(db, name)
    if rp is None:
        raise HTTPException(404, detail={"code": "PLUGIN_NOT_FOUND", "message": f"插件 {name} 不存在"})
    n = await feature_service.bulk_set_enabled(db, body.account_ids, name, enabled=True)
    await db.commit()
    return {"ok": True, "name": name, "applied": n}


@router.post("/{name}/disable-accounts")
async def api_disable_accounts(name: str, body: AccountPluginAction, db: DBSession, _user: CurrentUser):
    """按账号禁用远程插件。"""
    rp = await svc.get_by_name(db, name)
    if rp is None:
        raise HTTPException(404, detail={"code": "PLUGIN_NOT_FOUND", "message": f"插件 {name} 不存在"})
    n = await feature_service.bulk_set_enabled(db, body.account_ids, name, enabled=False)
    await db.commit()
    return {"ok": True, "name": name, "applied": n}


@router.post("/{name}/update", response_model=RemotePluginOut)
async def api_update(name: str, db: DBSession, _user: CurrentUser):
    """从远程仓库 git pull 并更新插件元数据。"""
    try:
        row = await svc.update(db, name)
        await db.commit()
        await db.refresh(row)
        await svc.trigger_reload(db, row.name)
        return row
    except RemotePluginNotFound as e:
        raise HTTPException(404, detail={"code": e.code, "message": e.message}) from e
    except (GitOperationFailed, InvalidPluginMetadata) as e:
        raise HTTPException(400, detail={"code": e.code, "message": e.message}) from e
    except RemotePluginError as e:
        raise HTTPException(400, detail={"code": e.code, "message": e.message}) from e


@router.delete("/{name}")
async def api_uninstall(name: str, db: DBSession, _user: CurrentUser):
    """卸载并删除指定远程插件（同时清理 Feature/AccountFeature 行）。"""
    found = await svc.uninstall(db, name)
    await db.commit()
    if found:
        await svc.trigger_reload(db, name)
    if not found:
        raise HTTPException(
            404, detail={"code": "PLUGIN_NOT_FOUND", "message": f"插件 {name} 不存在"}
        )
    return {"ok": True, "name": name}

File: frontend/src/api/remotePlugin.ts
import { api } from "@/lib/api";
import type { RemotePlugin, InstallRequest, AccountPluginAction } from "@/types/remotePlugin";

const BASE = "/api/remote-plugins";

export async function fetchRemotePlugins(): Promise<RemotePlugin[]> {
  const { data } = await api.get<RemotePlugin[]>(BASE);
  return data;
}

export async function installRemotePlugin(
  body: InstallRequest
): Promise<RemotePlugin> {
  const { data } = await api.post<RemotePlugin>(`${BASE}/install`, body);
  return data;
}

export async function enableRemotePlugin(
  name: string
): Promise<{ ok: boolean; name: string; enabled: boolean }> {
  const { data } = await api.post(`${BASE}/${encodeURIComponent(name)}/enable`);
  return data;
}

export async function disableRemotePlugin(
  name: string
): Promise<{ ok: boolean; name: string; enabled: boolean }> {
  const { data } = await api.post(
    `${BASE}/${encodeURIComponent(name)}/disable`
  );
  return data;
}

export async function enableRemotePluginForAccounts(
  name: string,
  body: AccountPluginAction
): Promise<{ ok: boolean; name: string; applied: number }> {
  const { data } = await api.post(
    `${BASE}/${encodeURIComponent(name)}/enable-accounts`,
    body
  );
  return data;
}

export async function disableRemotePluginForAccounts(
  name: string,
  body: AccountPluginAction
): Promise<{ ok: boolean; name: string; applied: number }> {
  const { data } = await api.post(
    `${BASE}/${encodeURIComponent(name)}/disable-accounts`,
    body
  );
  return data;
}

export async function updateRemotePlugin(name: string): Promise<RemotePlugin> {
  const { data } = await api.post<RemotePlugin>(
    `${BASE}/${encodeURIComponent(name)}/update`
  );
  return data;
}

export async function uninstallRemotePlugin(
  name: string
): Promise<{ ok: boolean; name: string }> {
  const { data } = await api.delete(
    `${BASE}/${encodeURIComponent(name)}`
  );
  return data;
}
